pisek/visualize.py

```diff
@@ -13,304 +9,330 @@
-) -> float:
-    points = 0.0
-    for subtask_id, sub_results in results.items():
-        points += evaluate_subtask(sub_results, config.subtasks[subtask_id].points)
-    return points
+def limit_result(result: LoggedResult, limit: float) -> LoggedResult:
+    if result.time <= limit and result.verdict == Verdict.timeout:
+        new_verdict = result.original_verdict
+        if new_verdict == Verdict.timeout:
+            new_verdict = Verdict.ok
 
+        return LoggedResult(
+            new_verdict, 1.0, result.time, result.test, result.original_verdict
+        )
+    if result.time > limit:
+        return LoggedResult(
+            Verdict.timeout, 0.0, result.time, result.test, result.original_verdict
+        )
+    return result
+
+
+class SolutionResults:
+    def __init__(
+        self, name: str, config: TaskConfig, results: list[LoggedResult]
+    ) -> None:
+        self.name = name
+        self._solution = config.solutions[name]
+        self._config = config
+        self._results = results
+
+        self._results.sort(key=lambda r: r.test)
+        self._results.sort(key=lambda r: r.time)
+        # self._results.sort(key=lambda r: r.verdict)
+        self._results.sort(key=self._get_subtask)
+
+    def _get_subtask(self, result: LoggedResult) -> int:
+        return min(
+            i
+            for i, sub in self._config.subtasks.items()
+            if sub.new_in_subtask(result.test)
+        )
 
-def evaluate_subtask(subtask_results: list[TestCaseResult], max_points):
-    return max_points * min(map(lambda r: r.points, subtask_results), default=0)
+    def _evaluate_results(
+        self, results: list[LoggedResult], expected: str
+    ) -> tuple[bool, bool, Optional[LoggedResult]]:
+        ok, definitive, breaker = self._config.evaluate_verdicts(
+            list(map(lambda r: r.verdict, results)), expected
+        )
+        return ok, definitive, results[breaker] if breaker is not None else None
+
+    @staticmethod
+    def from_log(
+        name: str, config: TaskConfig, testing_log, limit: float
+    ) -> "SolutionResults":
+        if name not in testing_log:
+            raise MissingSolution(name)
+
+        results = []
+        for result in testing_log[name]["results"]:
+            results.append(
+                limit_result(
+                    LoggedResult(
+                        Verdict[result["result"]],
+                        result["points"],
+                        result["time"],
+                        result["test"],
+                        Verdict[result["result"]],
+                    ),
+                    limit,
+                )
+            )
 
+        return SolutionResults(name, config, results)
+
+    def get_all(self) -> list[LoggedResult]:
+        return self._results
+
+    def get_by_subtask(self) -> dict[int, list[LoggedResult]]:
+        by_subtask: dict[int, list[LoggedResult]] = {
+            num: [] for num in self._config.subtasks
+        }
```

Comment:
Couldn't this be a list?

Reply:
No. Same item can appear in different subtasks.

Reply:
An item can appear in a nested list multiple times.
Since this is a dictionary that uses numbers from `0` to `n-1` as keys, it could be a `list[list[LoggedResult]]` instead.
It is true that this matches how subtasks are stored inside of `TaskConfig`, so it might be best to keep it as is.

Reply:
Ok, changed. I don't think it matters that much and we can change it later.